File: server/lib/rooms.js
export const rooms = new Map();

export const ROOM_CHARS = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';

export function generateRoomCode() {
  let code;
  do {
    code = Array.from({ length: 6 }, () => ROOM_CHARS[Math.floor(Math.random() * ROOM_CHARS.length)]).join('');
  } while (rooms.has(code));
  return code;
}

export function getCurrentTimestamp(room) {
  if (!room?.currentTrack) return 0;
  if (!room.isPlaying || !room.playbackState?.startedAt) {
    return room.playbackState?.timestamp || 0;
  }

  return room.playbackState.timestamp + (Date.now() - room.playbackState.startedAt) / 1000;
}

export function normalizeTrack(track) {
  if (!track) return null;

  const id = String(track.id || track.jamendoId || track.trackId || Date.now());
  const streamUrl = track.streamUrl || track.audio || track.url || track.fallbackUrl || '';

  return {
    id,
    title: track.title || track.name || 'Unknown Track',
    artistName: track.artistName || track.artist_name || '',
    albumName: track.albumName || track.album_name || '',
    thumbnail: track.thumbnail || track.image || track.album_image || '',
    duration: Number(track.duration || 0),
    streamUrl,
    audioDownloadUrl: track.audioDownloadUrl || track.audiodownload || '',
    shareUrl: track.shareUrl || track.shareurl || '',
    licenseUrl: track.licenseUrl || track.license_ccurl || '',
    source: track.source || 'jamendo',
    jamendoId: String(track.jamendoId || track.id || ''),
    isYouTube: Boolean(track.isYouTube),
    videoId: track.videoId || null,
  };
}

export function createRoomState({ roomId, user }) {
  return {
    id: roomId,
    users: new Map([[user.id, user]]),
    hostId: user.id,
    currentTrack: null,
    queue: [],
    history: [],
    isPlaying: false,
    repeatMode: 'off',
    shuffleMode: false,
    volume: 70,
    playbackState: { timestamp: 0, startedAt: null },
    lastMutationAt: Date.now(),
  };
}

export function serializeRoom(room) {
  if (!room) return null;

  const timestamp = getCurrentTimestamp(room);
  return {
    id: room.id,
    hostId: room.hostId,
    users: Array.from(room.users.values()),
    currentTrack: room.currentTrack,
    queue: room.queue,
    history: room.history,
    isPlaying: room.isPlaying,
    repeatMode: room.repeatMode,
    shuffleMode: room.shuffleMode,
    volume: room.volume,
    masterTimestamp: timestamp,
    lastSyncTime: room.playbackState?.startedAt || room.lastMutationAt || Date.now(),
  };
}

function isDuplicateTrack(queue, track) {
  return queue.some((item) => item.id === track.id || item.streamUrl === track.streamUrl);
}

export function addToQueue(room, track) {
  const normalized = normalizeTrack(track);
  if (!room || !normalized) return null;
  if (isDuplicateTrack(room.queue, normalized) || room.currentTrack?.id === normalized.id) return normalized;

  room.queue.push(normalized);
  room.lastMutationAt = Date.now();
  return normalized;
}

export function removeFromQueue(room, trackId) {
  if (!room || !trackId) return null;
  const index = room.queue.findIndex((track) => track.id === trackId);
  if (index < 0) return null;
  const [removed] = room.queue.splice(index, 1);
  room.lastMutationAt = Date.now();
  return removed;
}

export function clearQueue(room) {
  if (!room) return;
  room.queue = [];
  room.lastMutationAt = Date.now();
}

export function setCurrentTrack(room, track, { autoplay = true } = {}) {
  if (!room) return null;
  const normalized = normalizeTrack(track);
  if (!normalized) return null;

  if (room.currentTrack) {
    room.history = [room.currentTrack, ...room.history].slice(0, 50);
  }

  room.currentTrack = normalized;
  room.isPlaying = autoplay;
  room.playbackState = {
    timestamp: 0,
    startedAt: autoplay ? Date.now() : null,
  };
  room.lastMutationAt = Date.now();
  return normalized;
}

export function updatePlayback(room, { isPlaying, timestamp, startedAt = null }) {
  if (!room) return null;

  const now = Date.now();
  room.isPlaying = Boolean(isPlaying);
  room.playbackState.timestamp = Number(timestamp || 0);
  room.playbackState.startedAt = room.isPlaying ? startedAt || now : null;
  room.lastMutationAt = now;
  return getCurrentTimestamp(room);
}

export function moveToPreviousTrack(room) {
  if (!room) return null;

  const previous = room.history.shift();
  if (!previous) return null;

  if (room.currentTrack) {
    room.queue.unshift(room.currentTrack);
  }

  room.currentTrack = previous;
  room.isPlaying = true;
  room.playbackState = { timestamp: 0, startedAt: Date.now() };
  room.lastMutationAt = Date.now();
  return previous;
}

export function advanceTrack(room) {
  if (!room) return null;

  let nextTrack = null;
  if (room.shuffleMode && room.queue.length > 1) {
    const index = Math.floor(Math.random() * room.queue.length);
    nextTrack = room.queue.splice(index, 1)[0];
  } else {
    nextTrack = room.queue.shift() || null;
  }

  if (!nextTrack) {
    if (room.repeatMode === 'one' && room.currentTrack) {
      room.playbackState = { timestamp: 0, startedAt: Date.now() };
      room.isPlaying = true;
      room.lastMutationAt = Date.now();
      return room.currentTrack;
    }

    if (room.repeatMode === 'all' && room.history.length > 0) {
      nextTrack = room.history.pop();
    } else {
      room.isPlaying = false;
      room.playbackState = { timestamp: 0, startedAt: null };
      room.lastMutationAt = Date.now();
      return null;
    }
  }

  if (room.currentTrack) {
    room.history = [room.currentTrack, ...room.history].slice(0, 50);
  }

  room.currentTrack = nextTrack;
  room.isPlaying = true;
  room.playbackState = { timestamp: 0, startedAt: Date.now() };
  room.lastMutationAt = Date.now();
  return nextTrack;
}

File: server/app/rooms.py
import secrets
import string
import time
from typing import Optional, Dict
from .models import Room, User, Track, QueueItem, PlaybackState, RepeatMode

rooms: Dict[str, Room] = {}

CODE_CHARS = string.ascii_uppercase.replace("I", "").replace("O", "") + string.digits.replace("0", "").replace("1", "")


def generate_code() -> str:
    while True:
        code = "".join(secrets.choice(CODE_CHARS) for _ in range(6))
        if code not in rooms:
            return code


def create_room(host_name: str) -> Room:
    code = generate_code()
    user_id = f"user_{secrets.token_hex(4)}"
    now = time.time()
    room = Room(
        id=code,
        name=f"Room {code}",
        host_id=user_id,
        users=[User(id=user_id, name=host_name, is_host=True, joined_at=now)],
        playback=PlaybackState(last_updated=now),
        created_at=now,
    )
    rooms[code] = room
    return room, user_id


def get_room(code: str) -> Optional[Room]:
    return rooms.get(code.upper())


def join_room(code: str, user_name: str) -> tuple[Optional[Room], Optional[str], Optional[str]]:
    room = get_room(code)
    if not room:
        return None, None, "Room not found"
    user_id = f"user_{secrets.token_hex(4)}"
    now = time.time()
    room.users.append(User(id=user_id, name=user_name, is_host=False, joined_at=now))
    return room, user_id, None


def leave_room(code: str, user_id: str) -> Optional[Room]:
    room = get_room(code)
    if not room:
        return None
    room.users = [u for u in room.users if u.id != user_id]
    if not room.users:
        del rooms[code.upper()]
        return None
    if room.host_id == user_id and room.users:
        room.host_id = room.users[0].id
        room.users[0].is_host = True
    return room


def add_to_queue(room: Room, track: Track, user_id: str) -> QueueItem:
    item = QueueItem(track=track, added_by=user_id, added_at=time.time())
    room.queue.append(item)
    return item


def remove_from_queue(room: Room, track_id: str) -> bool:
    before = len(room.queue)
    room.queue = [q for q in room.queue if q.track.id != track_id]
    return len(room.queue) < before


def clear_queue(room: Room):
    room.queue.clear()


def set_track(room: Room, track: Track) -> Track:
    if room.playback.track:
        room.history.append(room.playback.track)
        if len(room.history) > 50:
            room.history = room.history[-50:]
    now = time.time()
    room.playback.track = track
    room.playback.position = 0
    room.playback.is_playing = True
    room.playback.started_at = now
    room.playback.last_updated = now
    room.queue = [q for q in room.queue if q.track.id != track.id]
    return track


def advance_track(room: Room) -> Optional[Track]:
    if room.queue:
        next_item = room.queue.pop(0)
        return set_track(room, next_item.track)
    if room.repeat_mode == RepeatMode.one and room.playback.track:
        return set_track(room, room.playback.track)
    if room.repeat_mode == RepeatMode.all and room.history:
        prev = room.history.pop(0)
        return set_track(room, prev)
    room.playback.is_playing = False
    room.playback.started_at = None
    room.playback.last_updated = time.time()
    return None


def previous_track(room: Room) -> Optional[Track]:
    if room.playback.track:
        room.queue.insert(0, QueueItem(track=room.playback.track, added_by=room.host_id))
    if room.history:
        prev = room.history.pop()
        return set_track(room, prev)
    room.playback.position = 0
    room.playback.started_at = time.time()
    room.playback.last_updated = time.time()
    return room.playback.track


def get_sync_state(room: Room) -> dict:
    pos = room.playback.position
    if room.playback.is_playing and room.playback.started_at:
        elapsed = time.time() - room.playback.started_at
        pos = room.playback.position + elapsed
    return {
        "track": room.playback.track.model_dump() if room.playback.track else None,
        "is_playing": room.playback.is_playing,
        "position": pos,
        "server_time": time.time(),
        "volume": room.volume,
        "repeat_mode": room.repeat_mode.value,
        "shuffle_mode": room.shuffle_mode,
    }
